File: src/seekerDemandLink.js
/** Link a kitchen commitment to the newest matching seeker demand (same area + menu). */
export async function enrichVendorBidWithSeekerDemand(
  record,
  seekerDemandStore
) {
  if (!seekerDemandStore?.listRecent || record.order_code) {
    return record;
  }
  const rows = await seekerDemandStore.listRecent({ limit: 100 });
  const locality = String(record.locality_key ?? "").trim();
  const offerId = String(record.standard_offer_id ?? "").trim();
  const match = rows.find(
    (row) =>
      String(row.locality_key ?? "").trim() === locality &&
      String(row.standard_offer_id ?? "").trim() === offerId &&
      row.order_code
  );
  if (!match) {
    return record;
  }
  return {
    ...record,
    seeker_demand_id: match.id,
    order_code: match.order_code,
    commitment_status: record.commitment_status ?? "committed"
  };
}

export function connectionNotifyRecipientIds({
  seekerDemand,
  kitchenCommitment,
  pledgeRecords
}) {
  const ids = new Set();
  if (seekerDemand?.reported_by_user_id) {
    ids.add(seekerDemand.reported_by_user_id);
  }
  if (kitchenCommitment?.submitted_by_user_id) {
    ids.add(kitchenCommitment.submitted_by_user_id);
  }
  for (const pledge of pledgeRecords) {
    if (pledge.pledged_by_user_id) {
      ids.add(pledge.pledged_by_user_id);
    }
  }
  return [...ids];
}

File: src/main/java/org/sharingbridge/integration/service/SeekerDemandLink.java
package org.sharingbridge.integration.service;

import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public final class SeekerDemandLink {

    private SeekerDemandLink() {}

    public static Map<String, Object> enrichVendorBidWithSeekerDemand(
            Map<String, Object> record, List<Map<String, Object>> recentSeekerDemands) {
        if (recentSeekerDemands == null || record.get("order_code") != null) {
            return record;
        }
        String locality = String.valueOf(record.get("locality_key") == null ? "" : record.get("locality_key")).trim();
        String offerId =
                String.valueOf(record.get("standard_offer_id") == null ? "" : record.get("standard_offer_id"))
                        .trim();
        for (Map<String, Object> row : recentSeekerDemands) {
            String rowLocality =
                    String.valueOf(row.get("locality_key") == null ? "" : row.get("locality_key")).trim();
            String rowOffer =
                    String.valueOf(row.get("standard_offer_id") == null ? "" : row.get("standard_offer_id"))
                            .trim();
            if (rowLocality.equals(locality) && rowOffer.equals(offerId) && row.get("order_code") != null) {
                Map<String, Object> next = JsValues.copy(record);
                next.put("seeker_demand_id", row.get("id"));
                next.put("order_code", row.get("order_code"));
                next.put(
                        "commitment_status",
                        record.get("commitment_status") != null
                                ? record.get("commitment_status")
                                : "committed");
                return next;
            }
        }
        return record;
    }

    public static List<String> connectionNotifyRecipientIds(
            Map<String, Object> seekerDemand,
            Map<String, Object> kitchenCommitment,
            List<Map<String, Object>> pledgeRecords) {
        Set<String> ids = new LinkedHashSet<>();
        if (seekerDemand != null && seekerDemand.get("reported_by_user_id") != null) {
            ids.add(String.valueOf(seekerDemand.get("reported_by_user_id")));
        }
        if (kitchenCommitment != null && kitchenCommitment.get("submitted_by_user_id") != null) {
            ids.add(String.valueOf(kitchenCommitment.get("submitted_by_user_id")));
        }
        if (pledgeRecords != null) {
            for (Map<String, Object> pledge : pledgeRecords) {
                if (pledge.get("pledged_by_user_id") != null) {
                    ids.add(String.valueOf(pledge.get("pledged_by_user_id")));
                }
            }
        }
        return new ArrayList<>(ids);
    }
}
